Initial commit

FutureSharks opened the pull request:
There is no code in this repo so I'm creating a PR.

I've started with what is in [this module](https://github.com/claranet/terraform-aws-aurora) and then made a couple changes:

- Auto generate the master password by default as it's more secure and means you don't need to have the password in your TF code.
- Better defaults for ports as currently a Postgres cluster defaults to using the MySQL port.

I will make more changes very soon.

Why not just commit to that other module? I think it's better to have it here in the `terraform-aws-modules` Github org as it will get more support and PRs from the community.

max-rocket-internet commented:
FYI @antonbabenko

FutureSharks commented on README.md:
```diff
@@ -1,5 +1,196 @@
 # terraform-aws-rds-aurora
-Terraform module which creates RDS Aurora resources on AWS
 
+Creates a AWS Aurora RDS cluster, instances, DB subnet group and optionally:
```
> As a separate idea, how hard would it be to extend https://github.com/terraform-aws-modules/terraform-aws-rds module to support Aurora specific features?

I considered this but the problem is that although Aurora falls under the RDS umbrella, the resources are actually quite different. It's not just a different engine for the existing RDS resource. It has a cluster and then instances that are members of the cluster. It can have a large amount of cluster instances/replicas to scale read load, up to 15. It can autoscale. It has a serverless option. I think that trying to combine it into that existing module not ideal, the only resources it would share is the option group and subnet group. Then there would need to be 3 extra resources that are solely for aurora.

> This resource module should not care about anything expect required RDS specific resources, so there is no need to bring SNS, Cloudwatch, Route53, App Autoscaling, etc.

OK I can remove SNS/cloudwatch if you don't want to include basic monitoring. I can also remove route53 record.

Autoscaling should stay. This isn't "App Autoscaling", this is [scaling of the replicas](https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/Aurora.Integrating.AutoScaling.html) and is one of the coolest features of Aurora.

antonbabenko replied:
Thanks for the detailed answer.

Yes, let's keep Aurora separately from the existing RDS module and include:
1. Monitoring (enhanced monitoring option in RDS, if there is such in Aurora)
2. Autoscaling

FutureSharks commented:
@antonbabenko OK I've made those changes.

In a broader context, what are your thoughts for optionally creating cloudwatch alarms in modules? I would like to make a case to include at least some basic alarms as an option in modules because:

1. Cloudwatch alarms are very tightly coupled to the resources they are for due to the `dimensions` and `statistic` arguments. Sure you could create them outside the module but I would argue it can quickly get quite messy/complicated and within a module is the perfect place to take care of this.

2. In a production environment cloudwatch alarms are not really optional because there are metrics that are very important that are ONLY available via cloudwatch. i.e. there is no other way to get that information. Some examples are `CPUCreditBalance` for t2 instances and things like `AuroraReplicaLag`. And for of course there are many others for other types of resources. Other important metrics like load, latency, IOPs, transactions etc are available using a third party monitoring system but are just so easy to monitor in cloudwatch as it's built in.

FutureSharks commented:
@antonbabenko can you re-review this PR?

antonbabenko commented:
I have just reviewed it and I have some comments which I feel should be addressed right from the start of the module.

PLEASE copy 100% from [terraform-aws-rds repo](https://github.com/terraform-aws-modules/terraform-aws-rds/), this includes:

- examples directory with several supported configurations
- documentation should not be generated manually (copy and use hooks from RDS module)
- remove parts of resources which don't belong to Aurora (as we discussed earlier. SNS, for example). Make working example to show how each of this feature/resource can be used, but make them 100% optional.

If you want I can make most of this and push, but unfortunately (as you, unfortunately, have already experienced) my response time varies a lot lately. I should have time during week-ends, if you want.

FutureSharks commented:
All good, I'll make those changes tomorrow.

FutureSharks commented:
OK @antonbabenko  I have:

- Added 3x examples. One MySQL, one PostgreSQL and one with autoscaling and enhanced monitoring.
- Documentation was already generated using `terraform-docs` but I have added your pre commit hook config also.
- I have removed the SNS topic used for monitoring from the example TF code. This wasn't in the module anyway.

Anything else?

antonbabenko commented:
Looks good. Let me merge this and apply few changes to be the same as RDS, VPC, etc in a separate PR.

FutureSharks commented:
Cool. Feel free to create some issues for the specific things you mean and I'll make PRs.
